File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
class Solution {
    private int width, length;
    
    private void move(String route, int[] cur, char[][] matrix) {
        String[] routeParts = route.split(" ");
        String op = routeParts[0];
        int n = Integer.parseInt(routeParts[1]);
        int dy = 0;
        int dx = 0;
        
        switch (op) {
            case "N" -> dy -= 1;
            case "S" -> dy += 1;
            case "W" -> dx -= 1;
            case "E" -> dx += 1;
        }
        
        int ny = cur[0] + dy * n;
        int nx = cur[1] + dx * n;
        
        if (ny < 0 || ny >= length || nx < 0 || nx >= width) return;
        
        for (int i = 1; i <= n; i++) {
            int y = cur[0] + dy * i;
            int x = cur[1] + dx * i;
            
            if (matrix[y][x] == 'X') return;
        }
        
        cur[0] = ny;
        cur[1] = nx;
    }
    
    public int[] solution(String[] park, String[] routes) {
        this.length = park.length;
        this.width = park[0].length();
        
        int[] cur = new int[2];
        char[][] matrix = new char[length][width];

        for (int i = 0; i < length; i++) {
            matrix[i] = park[i].toCharArray();
            for (int j = 0; j < width; j++) {
                if (matrix[i][j] == 'S') {
                    cur[0] = i;
                    cur[1] = j;   
                }
            }
        }
        
        for (String route : routes) {
            move(route, cur, matrix);
        }
        
        return cur;
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.kt
class Solution {
    private var width = 0
    private var length = 0
    
    private fun move(route: String, cur: IntArray, matrix: Array<CharArray>) {
        val (op, nStr) = route.split(" ")
        val n = nStr.toInt()
        var dy = 0
        var dx = 0
        
        when (op) {
            "N" -> dy -= 1
            "S" -> dy += 1
            "W" -> dx -= 1
            "E" -> dx += 1
        }
        
        val ny = cur[0] + dy * n
        val nx = cur[1] + dx * n
        
        if (ny !in 0 until length || nx !in 0 until width) return
        
        for (i in 1..n) {
            val y = cur[0] + dy * i
            val x = cur[1] + dx * i
            if (matrix[y][x] == 'X') return
        }
        
        cur[0] = ny
        cur[1] = nx
    }
    
    fun solution(park: Array<String>, routes: Array<String>): IntArray {
        length = park.size
        width = park[0].length
        
        var cur = IntArray(2)
        val matrix = park.map { it.toCharArray() }.toTypedArray()
        
        for (i in 0 until length) {
            for (j in 0 until width) {
                if (matrix[i][j] == 'S') {
                    cur[0] = i
                    cur[1] = j
                }
            }
        }
        
        for (route in routes) {
            move(route, cur, matrix)   
        }
        
        return cur
    }
}